### DISCUSSION TITLE
Versioning considerations

### DISCUSSION THREAD

**Discussion by dav1312** [official-stockfish/Stockfish#3967] [Ideas]
**Abandoning the use of compilation dates for versioning**

Currently Stockfish uses the compilation date to determine the version of the engine which causes multiple issues and confusion
<https://github.com/official-stockfish/Stockfish/blob/9f6bcb38c032a18c8b1aec318d5c7255827f9c7b/src/misc.cpp#L70-L72>

For example, the version that participated in the [CCC 17 Rapid Finals](https://www.chess.com/computer-chess-championship#event=ccc-17-rapid-finals) was `190322` but since the engine was compiled a few days later, the version turned out to be `220322` which doesn't really "exist".

This can cause other weird situations like compiling a version from a year ago returns the same version as compiling the latest one which doesn't make any sense.

This also causes problems when multiple commits are made on the same day because you can't really identify them using the version which again renders the version rather useless
e.g. https://discord.com/channels/435943710472011776/724661489381670923/948353154976714782
![image](https://user-images.githubusercontent.com/63931154/161593042-e90d0297-a1f0-44e2-bec8-c2981e4606a0.png)

This kind of versioning makes it very confusing for the average user too that might not understand what that very big number means
e.g. https://discord.com/channels/435943710472011776/733545871911813221/959838011376406649
![image](https://user-images.githubusercontent.com/63931154/161593731-5df7b516-477f-458b-8b91-4b4eec9f315a.png)

So my idea would be to do something similar to [Koivisto](https://github.com/Luecx/Koivisto/commit/f6b96c79e7c9fb452aafbffa9bf7afbf4aba05b9), increase the version on each commit, but since the mayor and minor versions already have a purpose (<https://github.com/official-stockfish/Stockfish/discussions/3368>), the idea would be to increase the patch number (e.g. 14.1 -> 14.1.1 -> 14.1.2).
This should clear all confusion regarding versions because there could not be two versions with the same number, each commit would be completely identifiable and the average user should be much more familiarized with this versioning style.

**Comment by kenta2** [comment]
I suggest the first part of version string be hardcoded and bumped after each release (e.g., "14.1"), and the build process derives a parenthesized suffix based on git commit.  Another project does this in Makefile:

```
ifneq ("$(wildcard $(SRCDIR)/.git)","")
PKGVER := g$(shell git -C $(SRCDIR) describe --always --dirty --exclude '*')
else
PKGVER :=
endif
```

Thus, the complete version string might look like this: "14.1 (g012abcd)".

**Reply by dav1312** [reply]
While this would solve some of the problems it would still be rather confusing for the average user

**Comment by vondele** [comment]
I do agree that our current 'compile time date' version number is not very good. 

Now alternatives should be automatic (like not having to change the version string manually or with a script on each commit), and ideally work without git being installed (like a user might just compile a .tar.gz of the sources and not have git installed).

**Reply by dav1312** [reply]
Maybe there is a way of doing it with gh actions

**Reply by vondele** [reply]
this might be, something like a hook that increments a number in a file named VERSION upon commit to master ?

**Reply by dav1312** [reply]
It should be possible with some (probably complex) bash script
The script would also need to identify if the increment should be patch, minor or mayor
Maybe adding a string next to the version that the script could read?
The string would always be "patch" and it could be changed to "minor" or "mayor"

**Comment by vondele** [comment]
This got fixed with https://github.com/official-stockfish/Stockfish/commit/e048d1182562eb7ce1eae45f626681206da446b8

**Reply by dav1312** [reply]
2/3 🙃

No more confusion:
- [x] when Stockfish is compiled a few days before the commit is done.
- [x] when multiple commits are made on the same day.
- [ ] for the average user.

Maybe some day it will be 15.1.[47](https://github.com/official-stockfish/Stockfish/compare/758f9c9350abee36a5865ec701560db8ea62004d...d3860f8d5efa4d9df726fa605b24689d6a829c7e)-d3860f8d 😄
